tonconnect-server/TonConnect2.md

```diff
@@ -0,0 +1,533 @@
+        "url": "https://api.example.com/...",
+        "headers": {
+            "X-Push-ID": "f8a90d7edad893...",
+        },
+    },
+    "lang-iso639-1": "en" | "zh" | ... 
+}
+```
+
+### ton-addr
+
+User provides the address with the proof of ownership.
+
+If you store some user-specific data in the DB, then make sure to use [client ID](#client-id) as a reliable identifier.
+
+This feature is limited to the simple individual wallets: multisig/lockup and smart contract wallets are not supported.
+
+Request:
+
+```
+{
+    "type": "ton-addr",
+}
+```
+
+Response:
+
+```
+{
+    "type": "ton-addr",
+    "address": "EQrt...s7Ui",    
+    "pubkey": "Pub6...2k3y", // base64-encoded Ed25519 public key
+    "signature": "Gt562...g5s8D=", // base64-encoded ed25519 signature
+    "wallet_id": null | integer, // should be omitted in usual cases
+    "wallet_version": "v4R2", // supported values: "v3R1", "v3R2", "v4R1", "v4R2"
+}
+```
+
+To create/verify the signature, construct the following message for Ed25519 algorithm:
+
+```
+"tonconnect/ton-addr/" || <wallet_version> || "/" || <address> || "/" || <client_id>
+```
+
+where:
+
+* `wallet_version` is encoded in ASCII, verification fails for unsupported values,
+* `address` is encoded in ASCII (standard user-readable, as provided in the request params),
+* `client_id` is a 32-byte binary [Client ID](#client-id).
+
+The resulting signature is bound to the user's public key, the service (via Client ID) and the concrete wallet version.
+
+**Note:** the proof is bound to the [Client ID](#client-id) and can be replayed across multiple sessions authenticating with the same client ID. This is by design, since the supported addresses do not support key rotation: if the user were able to demonstrate ownership once, they would normally be able to do it going forward.
+
+Validation rules:
+
+1. Construct wallet contract using the `pubkey`, `wallet_id` and `wallet_version` (`wc` could be read from the `address`).
+2. Check that the wallet contract address is equal to `address`.
+3. Verify the `signature` using the `pubkey`.
+4. Return `address` as TON address with verified ownership.
+
+
+
+
+### AuthRequest
+
+AuthRequest is suitable for centralized services to confirm offchain actions via the wallet.
+
+A JSON-encoded object with the following layout:
+
+```
+{
+    "protocol": "ton-auth",
+    "v2": {
+       "body": (base64-encoded signed AuthRequestBody),
+       "pk": (base64-encoded public key),
+       "dns": "example.ton",
+    }
+}
+```
+
+
+### AuthRequestBody
+
+A JSON-encoded object with the following layout:
+
+```
+{
+    "type": "v2-auth-req",
+    "info": {...},
+    "payload": ..., // arbitrary data (string/object) for signing by the user.
+    "details": "...", // human-readable description of the action
+    "reply": {...},
+}
+```
+
+### AuthResponse
+
+AuthResponse contains [response body](#AuthResponseBody) encrypted towards [App Public Key](#app-keypair) using `NaCl.crypto_box`:
```

Comment:
If this flow mandates an app keypair, it forces the app to store the private key somewhere. This means that the app cannot be fully open-sourced and be decentralized.

Open sourced dapps should also be capable of requesting such offline signatures, so at the very least the encryption should be optional.